design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,148 @@
+1. An existing CephObjectStoreUser exists along with the accompanying Secret in the user's namespace.
+
+##### Use Case: Expose an Object Store Endpoint for Bucket Provisioning
+
+As an admin, I want to expose an existing object store to cluster users so they can begin bucket provisioning via the Kubernetes API._
+
+![admin-actions](./cobc-admin.png)
+
+1. The admin creates a StorageClass with `parameters[serviceName]=objectstore-service` and `parameters[serviceNamespace]=some-namespace`
+  - _Note:_ the `provisioner` field is ignored.  Defining the provisioner is unnecessary as Rook-Ceph is the only operator watching for CephObjectBucketClaims.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want to leverage the Kubernetes API to create Ceph Object Buckets. I expect to get back the bucket connection information in a Pod-attachable object._
+ 
+![user-actions](./cobc-user.png)
+
+0. The user creates a CephObjectBucketClaim in the user's namespace.
+1. The Rook Operator detects a new ObjectBucketClaim instance.  
+    1. The operator gets the user's credentials via `cephObjectBucketClaim.spec.secretName`.  
+    1. The operator gets the object store URL via `storageClass.parameters.serviceNamespace` + `storageClass.parameters.serviceName`.
+1. The operator uses the object store URL and credentials to make an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the CephObjectBucketClaim with relevant connection data for that bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+#####Use Case: Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
+
+1. The user deletes the CephObjectBucketClaim via `kubectl delete ...`.
+1. The CephObjectBucketClaim is marked for deletion and left in the foreground.
+1. The respective ConfigMap is deleted.
+1. The CephObjectBucketClaim is garbage collected.
+
+---
+
+## API Specifications
+
+### Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: cephobjectbucketclaims.rook.io
+spec:
+  group: ceph.rook.io
+  version: v1
+  scope: namespaced
+  names:
+      kind: CephObjectBucketClaim
+      listKind: CephObjectBucketClaimList
+      singular: cephobjectbucketclaim
+      plural: cephobjectbucketclaims
+```
+
+### Custom Resource
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectBucketClaim
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  labels:
+    rook.io/bucket-provisioner:
+    rook.io/object-bucket-claim: my-bucket-1 [1]
+spec:
+  storageClassName: some-object-store
+  secretName: my-s3-key-pair
+  bucketNamePrefix: prefix [2]
+```
+
+1. Added by the Rook-Ceph operator.
+1. The operator will append a hyphen followed by random characters to the string given here.
```

Comment:
how many random chars?

Reply:
I know the importance of supporting random bucket names. However, if we also support fixed bucket names then a `CephObjectBucketClaim` can work for an existing bucket too. Eg. if I can define the full bucket name then this API ends up supporting _Get-Or-Create_ semantics. If the named bucket already exists the resulting ConfigMap would contain the endpoint for that existing bucket.  I guess _if_ there's value in this idea it can always be added later...

Reply:
I can make an educated guess at the answer but I think @travisn might have a more informed opinion here.